### PR TITLE
8330532: Improve line-oriented text parsing in HotSpot

### PR DESCRIPTION
**Pull request by iklam**
(This PR is an alternative to https://github.com/openjdk/jdk/pull/18669 with a better API for reading lines of text)

HotSpot has a few cases where information is parsed from a file, or from a memory buffer, one line at a time. Example:

- https://github.com/openjdk/jdk/blob/064628471b83616b4463baa78618d1b7a66d0c7c/src/hotspot/share/cds/classListParser.cpp#L169
- https://github.com/openjdk/jdk/blob/064628471b83616b4463baa78618d1b7a66d0c7c/src/hotspot/share/compiler/compilerOracle.cpp#L1059-L1066

Common problems:
- They use a fixed buffer for reading a line, so long (but valid) lines will cause errors.
- There's ad-hoc code that deals with `FILE*` differently than from memory.

This RFE implements a common utility, `inputStream`, for reading lines from different sources of input (see `FileInput` and `MemoryInput`). We fixed only `ClassListParser` and `CompilerOracle` in this RFE, but we can fix other readers in follow-up RFEs.

The API allows other source of input to be implemented. For example, one could implement a `SocketInput` if there's a use case for it.

In the future, `inputStream` can be extended (or encapsulated in a higher-level reader class) to read typed input tokens (for example, integers, strings, etc.)

Credit:
The `inputStream` class and friends are contributed by @rose00 . See https://mail.openjdk.org/pipermail/hotspot-dev/2024-April/087077.html .

John's original version is in the draft PR https://github.com/openjdk/jdk/pull/18773. In order to minimize the size of this PR, I have kept only the functionalities for reading a line and a time. Other features, such as pushing back contents into the `inputStream`, could be added in follow-up PRs. (These removed features can be found in the commit history of this PR).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8330532](https://bugs.openjdk.org/browse/JDK-8330532): Improve line-oriented text parsing in HotSpot (**Enhancement** - P4)

### Reviewers
 * [Matias Saavedra Silva](https://openjdk.org/census#matsaave) (@matias9927 - **Reviewer**) ⚠️ Review applies to [b7e856a9](https://git.openjdk.org/jdk/pull/18833/files/b7e856a9897714097032704951fee10fbc1c71ca)
 * [Johan Sjölen](https://openjdk.org/census#jsjolen) (@jdksjolen - **Reviewer**) ⚠️ Review applies to [2ddbfea9](https://git.openjdk.org/jdk/pull/18833/files/2ddbfea9b7be397f9f6f500dbec856d7ccb1eb11)

### Contributors
 * John R Rose `<jrose@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18833/head:pull/18833` \
`$ git checkout pull/18833`

Update a local copy of the PR: \
`$ git checkout pull/18833` \
`$ git pull https://git.openjdk.org/jdk.git pull/18833/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18833`

View PR using the GUI difftool: \
`$ git pr show -t 18833`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18833.diff">https://git.openjdk.org/jdk/pull/18833.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18833#issuecomment-2062944145)

### CODE REVIEW

**Review by coleenp — COMMENTED** [review]
Just drive-by comments.

File: src/hotspot/share/utilities/istream.hpp [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+  NONCOPYABLE(inputStream);
+
+  static constexpr size_t SMALL_SIZE =  240 DEBUG_ONLY(*0 + 10);
+  static constexpr size_t BIG_SIZE   = 2048 DEBUG_ONLY(*0 + 20);
+
+ protected:
+  // Values for _input_state, to distinguish some phases of history:
+  // Do we need to read more input (NTR)?  Did we see EOF already?
+  // Was there an error getting input or allocating buffer space?
+  enum IState { NTR_STATE, EOF_STATE, ERR_STATE };
```
Comment (coleenp) [inline src/hotspot/share/utilities/istream.hpp:92]: Enum class is preferable.

File: src/hotspot/share/utilities/istream.hpp [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+    : _fs(fs)
+  { }
+
+ public:
+  // just forward all the constructor arguments to the wrapped line-input class
+  template<typename... Arg>
+  FileInput(Arg... arg)
+    : _fs(_private_fs), _private_fs(arg...)
+  {
+  }
```
Comment (coleenp) [inline src/hotspot/share/utilities/istream.hpp:374]: can you put  {} on line 372, this doesn't need so much vertical space.

Comment (coleenp) [reply src/hotspot/share/utilities/istream.hpp:374]: Is this ... in template typename parameters allowed by our coding standard (whatever it means?)

Comment (rose00) [reply src/hotspot/share/utilities/istream.hpp:374]: Yes.  It’s called “variadic templates”; see `doc/hotspot-style.md`.   The usage appears twice already, in `metaprogramming/logical.hpp` and `asm/register.hpp`.  The usage here is pretty straightforward IMO:  It just forwards constructor argument lists unchanged.  This is useful for wrappers (where you construct the wrapped thing as a field) or subclass factoring (where you construct the super).

File: src/hotspot/share/utilities/istream.hpp [hunk 3]
```diff
@@ -0,0 +1,447 @@
...
+  // The purpose for the small buffer managed here is to store whole lines,
+  // and perhaps edit them in-place.
+  void prepare_to_fill_buffer(size_t& fill_offset, size_t& fill_length);
+
+  // Quick check for an initially incomplete buffer...
+  void preload() const {
+    if (need_to_read()) {
+      const_cast<inputStream*>(this)->fill_buffer();
+    }
+  }
```
Comment (jdksjolen) [inline src/hotspot/share/utilities/istream.hpp:207]: Why `const_cast` and assign this method `const` when it clearly is not? This shouldn't be `const`, is my point.

File: test/hotspot/gtest/utilities/test_istream.cpp [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+  for (int ncols = 0; ncols <= patlen; ncols++) {
+    if (SHORT_TEST) {
+      if (ncols < SHORT_NCOLS)  ncols = SHORT_NCOLS;
+      if (ncols > SHORT_NCOLS)  break;
+    } else if (ncols > COLS && ncols < patlen - COLS) {
+      ncols += ncols / 7;
+      if (ncols > patlen - COLS)  ncols = (patlen - COLS);
+    }
+    for (int lelen = 1; lelen <= 2; lelen++) {  // try both kinds of newline
+      //if (ncols > 0) ncols = (ncols == 1) ? (2*patlen)/3 : patlen;
```
Comment (matias9927) [inline test/hotspot/gtest/utilities/test_istream.cpp:290]: Leftover comment?

**Review by matias9927 — APPROVED** [review]
I have minor comments and considerations but otherwise, LGTM!

**Review by jdksjolen — COMMENTED** [review]
Hi,

The general concept of an `inputStream` is very good and I think this is useful, so I'm very happy that this is being added to Hotspot.

There's a design issue that I would like to be fixed, however. Consider the interface that `inputStream::Input` exposes, specifically the `close` operation:

```c++
    // Rewind so that the position appears to be the given one.
    // Return the new position, or else (size_t)-1 if the request fails.
    virtual size_t set_position(size_t position) { return -1; }

    // If it is backed by a resource that needs closing, do so.
    virtual void close() { }
```

This function:

```c++
void inputStream::set_input(inputStream::Input* input) {
  clear_buffer();
  if (_input != nullptr && _input != input) {
    _input->close();
  }
  _input = input;
  _input_state = NTR_STATE;
}
```

and finally this:

```c++
 // class inputStream
  Input* _input;   // where the input comes from or else nullptr
```

I'd like to see `set_input` be removed, as instead of changing the inputStream a new can be created (and replace the old one, if need be). Second, I'd like to see the `close` method to be removed as `inputStream` no longer has any code calling it.  Third, I'd like to see the `_input` field to go from being a pointer to being a reference as there is no point in an `inputStream` not having an input, and that input should not change.

Closing the resource is the responsibility of the owner, and this preferably happens in its destructor. If an explicit close method is required, then the owner will know what that method is, as the input object will be "concrete" (such as `FileInput`). In `MemoryInput` `close()`:ing the resource ought to be the same as `free()`:ing it, but here that is ignored and `close()` seems unimplemented?

Consider the following code, I'd be very surprised at this:

```c++
{
  FileInput fi("myfile");
  {
    inputStream(&fi);
    // ... Let's say I figure out that the file has at least 8 bytes
  }
  char bytearr[8];
  fi.set_position(0);
  int read_bytes = fi.read(&bytearr, 8);
  // read_bytes is 0 !??!
}
```

More generally, the class `inputStream::Input` should exist to satisfy the needs of `inputStream` and **not** as a general interface for any and all inputs. Any optional method cannot be depended upon anyway, so two paths will have to be written if any function takes an `inputStream::Input&` as input. I'd like to call upon "YAGNI" and "KISS" here, as I believe we're over complicating it without any actual need for the complexity.

File: src/hotspot/share/utilities/istream.hpp (hunk 3 above)
Comment (rose00) [reply src/hotspot/share/utilities/istream.hpp:207]: The method is const because the logical state of the stream is invariant, as visible to the API user.  If the implementation needs an invisible internal state change, it needs a const-cast (or mutable field, sometimes, will work).  If `preload` were made non-const as you suggest, we’d need to move the const-casting elsewhere, and it would be less clear that `preload` preserves API-visible state.  So the code, as it is, is the most convenient place to put the const-cast, as an internal implementation decision.

File: src/hotspot/share/utilities/istream.hpp [hunk 5]
```diff
@@ -0,0 +1,447 @@
...
+  // when next is called.
+
+  bool is_sane() const {
+    assert((_buffer == nullptr) == (_buffer_size == 0), "");
+    assert(_content_end <= _buffer_size, "");
+    assert(_beg <= _end && _end <= _content_end, "");
+    assert(_end <= _next && _next <= _content_end + NEXT_PHANTOM, "");
+    assert(_buffer_size == 0 || _next <= _buffer_size, "");
+    return true;
+  }
```
[the remark this answers is not in the document]
Comment (rose00) [reply src/hotspot/share/utilities/istream.hpp:150]: No, that’s not necessary.  There are many, many empty assert strings in HotSpot.  If there’s no message, it means “check the code logic here”.  You don’t need to say “invariant” or “sanity” or “must be” as a redundant means of conveying that message.  Although, some authors do this.  But if there are a long string of asserts, saying “invariant” that many times is simply noise.

File: src/hotspot/share/utilities/istream.hpp [hunk 6]
```diff
@@ -0,0 +1,447 @@
...
+  char   _line_ending;  // one of {0,1,2} for "", "\n", "\r\n"
+  char*  _buffer;       // scratch buffer holding at least the current line
+  size_t _buffer_size;  // allocated size of buffer
+  size_t _content_end;  // offset to end of valid contents of buffer
+  size_t _beg;          // offset in buffer to start of current line
+  size_t _end;          // offset to end of known current line (else content_end)
+  size_t _next;         // offset to known start of next line (else =end)
+  void*  _must_free;    // unless null, a malloc pointer which we must free
+  size_t _line_count;   // increasing non-resettable count of lines read
+  char   _small_buffer[SMALL_SIZE];  // buffer for holding lines
```
[the remark this answers is not in the document]
Comment (rose00) [reply src/hotspot/share/utilities/istream.hpp:108]: No, it’s the small buffer that is the initial estimate of the line buffer, which in general must grow by heap allocation.

(BTW, the presence of small_buffer is the reason your other suggestion about `set_input` is wrong; will explain there…)

File: src/hotspot/share/utilities/istream.hpp [hunk 7]
```diff
@@ -0,0 +1,447 @@
...
+    set_input(input);
+  }
+
+  virtual ~inputStream() {
+    if (_must_free)         handle_free();
+    if (_input != nullptr)  set_input(nullptr);
+  }
+
+  // Discards any previous input and sets the given input source.
+  void set_input(Input* input);
```
[the remark this answers is not in the document]
Comment (rose00) [reply src/hotspot/share/utilities/istream.hpp:249]: There may be a better way to design the way streams and inputs connect with each other, but removing `set_input` is a step in the wrong direction.  The reason is that a stream is designed to be stack-allocated, while an input source is something that can have a non-stack lifetime.

It is supposed to be cheap to create an input stream on stack, and part of the cost model comes from the small buffer, which is also stack allocated, and defers the need for heap allocation.  This means it could be a performance problem to “make a new one” as suggested above.  It’s better to allow the stack allocated one to modify its input source, if the input is either (a) multiplexed from different sources, or (b) determined AFTER the input stream is declared.

Yes, the normal use is to determine the input source first and then wrap the i-stream around it.  But if this fails, it is not the right answer to push the i-stream onto the heap; i-streams are strongly associated with the blocks of code that exercise them.  The right answer, when the input sources are “hard to manage”, is `set_input`.

File: src/hotspot/share/cds/classListParser.cpp [hunk 8]
```diff
@@ -450,6 +433,25 @@ void ClassListParser::error(const char* msg, ...) {
   va_end(ap);
 }
 
+void ClassListParser::check_class_name(const char* class_name) {
```
[the remark this answers is not in the document]
Comment (iklam) [reply src/hotspot/share/cds/classListParser.cpp:436]: `SystemDictionary::class_name_symbol()` assumes that the input is UTF8, but I cannot make that assumption here, as the `class_name` comes from a text file that can have arbitrary content.

File: src/hotspot/share/utilities/istream.hpp (hunk 1 above)
Comment (iklam) [reply src/hotspot/share/utilities/istream.hpp:92]: Is there an adopted style for the enumerators? I couldn't find it in the hotspot style guide. There's quite a variation today. E.g..

```
enum class vmSymbolID : int {  NO_SID = 0,   ....};
enum class DefaultsLookupMode { find, skip };
```

File: src/hotspot/share/utilities/istream.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (iklam) [reply src/hotspot/share/utilities/istream.hpp:108]: I added comments:
```
  char   _small_buffer[SMALL_SIZE];  // stack-allocated buffer for holding lines;
                                     // will switch to C_HEAP allocation when necessary.
```

File: src/hotspot/share/utilities/istream.hpp (hunk 3 above)
Comment (iklam) [reply src/hotspot/share/utilities/istream.hpp:207]: This page suggests using the `mutable` keyword for this situation (but `const_cast<>` is acceptable as a last resort)
https://isocpp.org/wiki/faq/const-correctness#mutable-data-members

However, to use `mutable` in this case, we would probably need to declare every field as `mutable`, this means that the meaning of `const` will be very difficult to understand (at least we can't use the C++ compiler to tell us which function can't (should) be `const` and which function cannot be `const`.)

Comment (iklam) [reply src/hotspot/share/utilities/istream.hpp:207]: On a separate note, `preload()` can cause data to be read from the input. For non-seekable input (such as sockets), this doesn't seem like a `const` operation to me.

Comment (rose00) [reply src/hotspot/share/utilities/istream.hpp:208]: I could be wrong about this, but it seems like it’s completely up to the C++ class to define what aspects of its implementation are fixed for its `const` methods.  The `const` keyword does not mean “everything is immutable about this, at all levels and inside all boundaries”.  But then it’s up to the class author to choose which boundary `const` applies to.  The thing wrapped inside the i-stream is, I will claim, implementation which can change state even in `const` functions.

If not, the alternative is to have almost no `const` functions at all in the i-stream class.  That’s less useful, because `const` means something very useful, in the context of the i-stream class.  It means that the line will not shift.  You can keep on reading, and even writing, the line buffer, as long as you call only `const` functions.

The reuse of the line buffer, in this way, is a core part of the i-stream design, and part of its value proposition:  You don’t get a new allocation on every read-line op (as you would with Java).  But in order to draw bright lines around THAT very useful notion of invariance (during which line buffers are reusable), you NEED `preload` to be const, even if it does internal book keeping, even if it consults the input source at times.

In short, a rigid idea of `const` is inconsistent with the carefully balanced performance characteristics of this design.

File: src/hotspot/share/utilities/istream.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (jdksjolen) [reply src/hotspot/share/utilities/istream.hpp:151]: I was not aware of that, thanks! Let's skip this change then.

File: src/hotspot/share/utilities/istream.hpp [hunk 9]
```diff
@@ -0,0 +1,449 @@
...
+  // Values for _input_state, to distinguish some phases of history:
+  // Do we need to read more input (NTR)?  Did we see EOF already?
+  // Was there an error getting input or allocating buffer space?
+  enum class IState : int { NTR_STATE, EOF_STATE, ERR_STATE };
+
+  // Named offset for _next relative to _content_end, of phantom '\n'.
+  static const int NEXT_PHANTOM = 1;
+
+  Input* _input;   // where the input comes from or else nullptr
+  IState _input_state;  // one of {NTR,EOF,ERR}_STATE
```
Comment (jdksjolen) [inline src/hotspot/share/utilities/istream.hpp:98]: This comment not necessary as the type describes its valid state.

File: src/hotspot/share/utilities/istream.hpp [hunk 10]
```diff
@@ -0,0 +1,449 @@
...
+  char   _line_ending;  // one of {0,1,2} for "", "\n", "\r\n"
+  char*  _buffer;       // scratch buffer holding at least the current line
+  size_t _buffer_size;  // allocated size of buffer
+  size_t _content_end;  // offset to end of valid contents of buffer
+  size_t _beg;          // offset in buffer to start of current line
+  size_t _end;          // offset to end of known current line (else content_end)
+  size_t _next;         // offset to known start of next line (else =end)
+  void*  _must_free;    // unless null, a malloc pointer which we must free
```
Comment (jdksjolen) [inline src/hotspot/share/utilities/istream.hpp:106]: Reading this code, why do we set `_must_free` instead of simply having a method:

```c++
bool must_free() {
  return _buffer != &_small_buffer;
}
```

and just delete the `_must_free` field.

**Review by jdksjolen — COMMENTED** [review]
A couple of more comments.

And I implemented the simplification I suggested like this: https://github.com/jdksjolen/jdk/commit/b5bc0e945cb96c0a73a91981b937968e5cbac33c

Also, I use `override` instead of `virtual`, then we get compiler help if we for some reason (non-matching arg lists for example) don't actually override a super class's virtual function.

File: src/hotspot/share/utilities/istream.hpp (hunk 3 above)
Comment (jdksjolen) [reply src/hotspot/share/utilities/istream.hpp:208]: I had to read up on this and I believe that John is right.

To quote https://isocpp.org/wiki/faq/const-correctness#const-member-fns

>The trailing const on inspect() member function should be used to mean the method won’t change the object’s abstract (client-visible) state. That is slightly different from saying the method won’t change the “raw bits” of the object’s struct.

I'm not objecting to the usage of `const` for this method, considering this. Reviewing code is learning on the job!

File: src/hotspot/share/utilities/istream.hpp (hunk 10 above)
Comment (rose00) [reply src/hotspot/share/utilities/istream.hpp:106]: Good question.  There was a version of the code that accepted a user-supplied buffer, optionally.  In that case, `_must_free` was set false (or to a user-requested value), since it was up to the user whether the user-supplied buffer should be freed.  It could be a static buffer.  If there is no longer such an option in the existing constructors, then this field can be GC-ed.

File: src/hotspot/share/utilities/istream.hpp (hunk 9 above)
Comment (rose00) [reply src/hotspot/share/utilities/istream.hpp:98]: Sometimes redundant comments are helpful.  I think this one is.  YMMV.

Comment (rose00) [reply src/hotspot/share/utilities/istream.hpp:98]: The `override` keyword is nice; thank you.

I have already argued against the removal of `set_input`.  And `set_input` needs `close`.

I think `set_input` is not YAGNI but YIWNI = Yes I will need it.  The reply that “you can just wrap another i-stream around the new i-source” is fallacious because of the performance model of i-stream.

File: src/hotspot/share/utilities/istream.hpp (hunk 1 above)
Comment (rose00) [reply src/hotspot/share/utilities/istream.hpp]: According to my interpretation, the relevant style guide passage is about constants:

> * Constant names may be upper-case or mixed-case, according to historical necessity.  (Note: There are many examples of constants with lowercase names.)

Basically, “styles vary, follow precedent”.  I followed the precedent of `JavaThreadStatus` and others, as noted elsewhere in this PR.  OS-level constants have old-looking names like `EOF` and `FILENAME_MAX`.  That’s also the standard for Java constant names.

File: src/hotspot/share/utilities/istream.hpp (hunk 9 above)
Comment (jdksjolen) [reply src/hotspot/share/utilities/istream.hpp:98]: Sorry, I'm still not on board with the `close` operation and I'm against `set_input` calling `close()` :-). Why is it necessary for the `inputStream` to require a file to be re-opened if the `inputStream` switches from one file to another?

To be clear: OK, we want `set_input` because we don't want to allocate two small buffers, that's fine by me.

File: src/hotspot/share/utilities/istream.hpp (hunk 10 above)
Comment (iklam) [reply src/hotspot/share/utilities/istream.hpp]: I removed `_must_free` and added a new `has_c_heap_buffer()` method.

File: src/hotspot/share/utilities/istream.hpp (hunk 9 above)
Comment (iklam) [reply src/hotspot/share/utilities/istream.hpp:98]: @jdksjolen I have incorporated most of your suggested changes. I left this code in for now:

```
void inputStream::set_input(inputStream::Input* input) {
  clear_buffer();
  if (_input != nullptr && _input != input) {
    _input->close();
  }
  ...
```

I am also leaning towards removing the `close()` call. Otherwise it would be unsymmetrical - the `inputStream` doesn't open the `_input` automatically, but it will close it automatically for us.

It seems better to leave both the `open` and `close` to the caller the `inputStream`.

@rose00 what do you think?

File: src/hotspot/share/utilities/istream.hpp [hunk 11]
```diff
@@ -0,0 +1,396 @@
...
+  // buffered, if it is of normal size.
+  bool expand_buffer(size_t new_length);
+
+  // Make sure there is at least one line in the buffer, and set
+  // _beg/_end to indicate where it is.  Any content before _beg can
+  // be overwritten to make more room in the buffer.  If there is no
+  // more input, set the state up to indicate we are done.
+  bool fill_buffer();
+
+  // Find some room in the buffer so we call read on it.
```
Comment (jdksjolen) [inline src/hotspot/share/utilities/istream.hpp:196]: "so we **can** call read on it" probably

File: src/hotspot/share/utilities/istream.cpp [hunk 12]
```diff
@@ -0,0 +1,378 @@
...
+    _input->close();
+  }
+  _input = input;
+  _input_state = IState::NTR_STATE;
+}
+
+bool inputStream::fill_buffer() {
+  assert(!definitely_done(), "");  // caller responsibility
+  while (need_to_read()) {
+    size_t fill_offset, fill_length;
```
Comment (jdksjolen) [inline src/hotspot/share/utilities/istream.cpp:119]: Nit: It's fine to move these decls out of the loop.

File: src/hotspot/share/utilities/istream.hpp [hunk 13]
```diff
@@ -0,0 +1,396 @@
...
+    _input(nullptr),
+    _input_state(IState::NTR_STATE),
+    _line_ending(0),
+    _buffer(nullptr),
+    _buffer_size(0),
+    _content_end(0),
+    _beg(0),
+    _end(0),
+    _next(0),
+    _line_count(0) {}
```
Comment (jdksjolen) [inline src/hotspot/share/utilities/istream.hpp:236]: Explicitly initialize the `_small_buffer` (`0` it out?). Set `_buffer` and `_buffer_size` to the small buffer by default.

File: src/hotspot/share/utilities/istream.cpp [hunk 14]
```diff
@@ -0,0 +1,378 @@
...
+// the buffer are in use.
+void inputStream::prepare_to_fill_buffer(size_t& fill_offset,
+                                         size_t& fill_length) {
+  assert(need_to_read(), "");  // _next pointer out of the way
+  if (_buffer_size == 0) {
+    COV(PFB_X);
+    expand_buffer(sizeof(_small_buffer));
+    assert(_buffer_size > 0, "");
+    // and continue with at least a little buffer
+  }
```
Comment (jdksjolen) [inline src/hotspot/share/utilities/istream.cpp:173]: Get rid of this branch, small buffer now default.

File: src/hotspot/share/utilities/istream.cpp [hunk 15]
```diff
@@ -0,0 +1,378 @@
...
+}
+
+// Return true iff we expanded the buffer to the given length.
+bool inputStream::expand_buffer(size_t new_length) {
+  assert(new_length > _buffer_size, "");
+  char* new_buf = nullptr;
+  if (new_length <= sizeof(_small_buffer)) {
+    COV(EXB_S);
+    new_buf = &_small_buffer[0];
+    new_length = sizeof(_small_buffer);
```
Comment (jdksjolen) [inline src/hotspot/share/utilities/istream.cpp:273]: Remove this branch, cannot ever happen as default is the small buffer.

File: src/hotspot/share/utilities/istream.hpp [hunk 16]
```diff
@@ -0,0 +1,396 @@
...
+  size_t _content_end;  // offset to end of valid contents of buffer
+  size_t _beg;          // offset in buffer to start of current line
+  size_t _end;          // offset to end of known current line (else content_end)
+  size_t _next;         // offset to known start of next line (else =end)
+  size_t _line_count;   // increasing non-resettable count of lines read
+  char   _small_buffer[SMALL_SIZE];  // stack-allocated buffer for holding lines;
+                                     // will switch to C_HEAP allocation when necessary.
+
+  bool has_c_heap_buffer() {
+    return _buffer != nullptr && _buffer != &_small_buffer[0];
```
Comment (jdksjolen) [inline src/hotspot/share/utilities/istream.hpp:111]: No need to check for nullptr.

File: src/hotspot/share/utilities/istream.hpp [hunk 17]
```diff
@@ -0,0 +1,396 @@
...
+  // Take input from the given source.  Buffer only a modest amount.
+  inputStream(Input* input)
+    : inputStream()
+  {
+    set_input(input);
+  }
+
+  virtual ~inputStream() {
+    if (has_c_heap_buffer()) free_c_heap_buffer();
+    if (_input != nullptr)  set_input(nullptr);
```
Comment (jdksjolen) [inline src/hotspot/share/utilities/istream.hpp:247]: In case we remove close: remove this call.

**Review by jdksjolen — COMMENTED** [review]
Hi Ioi, thanks for looking at my changes. I've got some more changes :-). Mainly: Shouldn't the `_small_buffer` be assigned to the `_buffer` pointer by default? This simplifies the code a bit.

File: src/hotspot/share/utilities/istream.cpp [hunk 18]
```diff
@@ -0,0 +1,378 @@
...
+       COV_COUNT(casename) = 0;
+    DO_COV_CASES(CLEAR_COV_CASE)
+    #undef CLEAR_COV_CASE
+  }
+  cases  = num_cases;
+  total  = case_count;
+  zeroes = zero_count;
+  return old_mode;
+}
+#endif //ASSERT
```
[the remark this answers is not in the document]
Comment (jdksjolen) [reply src/hotspot/share/utilities/istream.cpp:378]: Key is to provide configure with `--enable-native-coverage`.

For `jib` we have `linux-x64-cov`, see https://github.com/openjdk/jdk/blob/master/make/conf/jib-profiles.js

File: src/hotspot/share/utilities/istream.hpp (hunk 13 above)
Comment (iklam) [reply src/hotspot/share/utilities/istream.hpp]: Fixed.

File: src/hotspot/share/utilities/istream.cpp (hunk 14 above)
Comment (iklam) [reply src/hotspot/share/utilities/istream.cpp]: Fixed.

**Comment by tstuefe** [comment]
Drive-by comment: do we really need to combine input and output capabilities in one class, fileStream? Why not a dedicated fileOutputStream? APIs like setPosition() do not make much sense on an output stream.

**Review by jdksjolen — APPROVED** [review]
Please apply my suggestions, we should use `override` rather than `virtual` when implementing virtual functions.

**Comment by jdksjolen** [comment]
Other than those two suggestions, the code looks good to me.

Thanks for your efforts in making this happen!

**Comment by iklam** [comment]
> Ioi, I would support reducing coupling with fileStream by removing old rewind and new position/remaining functions. But I?d rather keep the new functions, because I think they are likely to be useful. I have future uses in mind, which might or might not happen. For example, open a CDS archive or large config file, position the fileStream at the base address of some textual configuration data, and start reading.

Hi John, I decided to remove the position/set_position/remaining functions from the PR, as we no longer have test cases that cover them. When they are needed in the future, we can restore them from history and add the necessary tests.
